cartoframes/dataset.py

```diff
@@ -55,52 +58,61 @@ def __init__(self, table_name=None, schema='public', query=None, df=None, gdf=No
-            query=query, context=context, state=cls.STATE_REMOTE)
+    def from_query(cls, query, context):
+        return cls(query=query, context=context, state=cls.STATE_REMOTE)
 
     @classmethod
-    def from_dataframe(cls, df, table_name=None, schema='public', context=None):
-        return cls(
-            df=df, table_name=table_name, schema=schema, context=context, state=cls.STATE_LOCAL)
+    def from_dataframe(cls, df):
+        return cls(df=df, state=cls.STATE_LOCAL)
 
     @classmethod
-    def from_geodataframe(cls, gdf, table_name=None, schema='public', context=None):
-        return cls(
-            gdf=gdf, table_name=table_name, schema=schema, context=context, state=cls.STATE_LOCAL)
+    def from_geodataframe(cls, gdf):
+        return cls(gdf=gdf, state=cls.STATE_LOCAL)
 
     @classmethod
-    def from_geojson(cls, geojson, table_name=None, schema='public', context=None):
-        return cls(
-            gdf=load_geojson(geojson), table_name=table_name, schema=schema, context=context, state=cls.STATE_LOCAL)
+    def from_geojson(cls, geojson):
+        return cls(gdf=load_geojson(geojson), state=cls.STATE_LOCAL)
 
-    def upload(self, with_lonlat=None, if_exists='fail'):
-        if self.query and self.table_name is not None and not self.exists():
-            self.cc.batch_sql_client.create_and_wait_for_completion(
-                '''BEGIN; {drop}; {create}; {cartodbfy}; COMMIT;'''
-                .format(drop=self._drop_table_query(),
-                        create=self._create_table_from_query(self.query),
-                        cartodbfy=self._cartodbfy_query()))
-        else:
-            if self.df is None:
-                raise ValueError('You have to create a `Dataset` with a pandas.DataFrame to upload it to CARTO')
+    def upload(self, with_lonlat=None, if_exists='fail', table_name=None, schema=None, context=None):
+        if table_name:
+            self.table_name = normalize_name(table_name)
+        if schema:
+            self.schema = schema
+        if context:
+            self.cc = context
+
+        if self.table_name is None or self.cc is None:
+            raise CartoException('You should provide a table_name and context to upload data.')
+
+        if self.gdf is None and self.df is None and self.query is None:
+            raise CartoException('Nothing to upload.')
 
-            if not self.exists():
+        already_exists_error = NameError('Table with name {table_name} already exists in CARTO.'
+                                         ' Please choose a different `table_name` or use'
+                                         ' if_exists="replace" to overwrite it'.format(table_name=self.table_name))
+
+        # priority order: gdf, df, query
+        if self.gdf is not None:
+            warn('GeoDataFrame option is still under development. We will try the upload with DataFrame')
+
+        if self.df is not None:
+            if not self.exists() or if_exists == Dataset.REPLACE:
```

Comment:
let's rewrite as 

```suggestion
            if if_exists == Dataset.REPLACE or not self.exists():
```

and we can save a query to the database.